### DISCUSSION TITLE
Help needed to test the virtual mouse functionality

### DISCUSSION THREAD

**Discussion by nirenjan** [nirenjan/x52pro-linux#34] [General]
Hi everybody,

I have made a few changes to enable the virtual mouse controlled by the thumbstick on the throttle unit, and would like some help to test it out. Please checkout the `mouse-support` branch, and build it. There is a new dependency on `libevdev`, which can be installed on Ubuntu with `libevdev-dev` package, or on Arch with `libevdev` package.

If you don't have the dependency, then it will compile everything except the virtual mouse driver, and you can verify this by looking at the build logs. I've also increased the default logging verbosity in the systemd service file.

Please test this out and provide your feedback. Thank you in advance.

Tagging users who I believe have X52 devices and running Linux: @LittleCannon, @gabolander, @Daebis, @lbeltrame, @KishCom, @Xeraster, @colourful-land, @mauricioduarte01, @rogerxxxx, @winiciuscota

EDIT: Thanks to everybody who tested and reported issues. I've merged the virtual mouse functionality to the master branch, and should have a new release out within the week.

**Comment by lbeltrame** [comment]
Thanks for the heads up. I'll check this weekend.

**Reply by lbeltrame** [reply]
Tested it out today. I had to set speed to 4: movement is smooth but slower than what I'd expect (but I'm used to a high DPI mouse, so it might be just bias). Everything else seems to work fine.

**Reply by nirenjan** [reply]
Thank you for your help. Can you confirm which commit you were at when you tested the changes?

**Reply by lbeltrame** [reply]
I was at the tip of `mouse-support` branch, that is 06abc5636db6fbbb6b2ba62c12bb48a85278e0ec. 
I double checked on Windows to make sure it wasn't just the hardware doing that, and there it's much faster (but so fast it's hardly usable).

**Reply by nirenjan** [reply]
Thank you - I've pushed a change to f4905365de8f9f68ef758ea4c7e5a3556695452e, which should make the speed a lot more customizable. It now varies from 0 to 32 (inclusive), and should give you a better experience.

Please do let me know how it goes. Thanks

**Reply by lbeltrame** [reply]
This is much better. I've found my (personal) optimum at 15.

**Comment by LittleCannon** [comment]
Build it without problems.
Thumb-stick reacts, but refresh rate slow.
Mouse speed option it seems have no effect (ignore in journalctl 128 setting, I tried later with 5).

evtest (full couple of circles) and journactl in attachment
[evtest-mouse.txt](https://github.com/nirenjan/x52pro-linux/files/7170395/evtest-mouse.txt)
[journalctl_log.txt](https://github.com/nirenjan/x52pro-linux/files/7170398/journalctl_log.txt)

**Reply by nirenjan** [reply]
Can you try with speed 4? That is the maximum speed supported. Thanks.

**Reply by nirenjan** [reply]
Looks like a bug in the logging, it should say 0-4.

**Reply by LittleCannon** [reply]
Significantly better.
Also, pressure sensitivity seems to work quite OK.

**Reply by nirenjan** [reply]
I've changed the speed parameters, so it should be smoother even at speed 0. Can you give this new version a try? Thanks.

**Reply by LittleCannon** [reply]
You are right.
Much smoother and at speed 4 is almost like trackpoint.

**Reply by nirenjan** [reply]
Awesome, thanks for your help. I'll leave this branch for a little bit so that others can give it a test, and assuming no other issues, I'll merge it sometime next week.

**Reply by nirenjan** [reply]
I've changed the speed algorithm in f4905365de8f9f68ef758ea4c7e5a3556695452e, can you give this new version a try and let me know how it is? The speed varies from 0 - 32, speeds 1 - 4 are a little slower than the previous version, but the new speeds should make up for this. Thanks.

**Reply by LittleCannon** [reply]
Sorry for delay, somehow I missed your reply.
0-32 work perfectly. Personally, speed on 12 is perfect.

**Comment by Fr-Dae** [comment]
> Hi everybody,
> 
> I have made a few changes to enable the virtual mouse controlled by the thumbstick on the throttle unit, and would like some help to test it out. Please checkout the `mouse-support` branch, and build it. There is a new dependency on `libevdev`, which can be installed on Ubuntu with `libevdev-dev` package, or on Arch with `libevdev` package.
> 
> If you don't have the dependency, then it will compile everything except the virtual mouse driver, and you can verify this by looking at the build logs. I've also increased the default logging verbosity in the systemd service file.
> 
> Please test this out and provide your feedback. Thank you in advance.
> 
> Tagging users who I believe have X52 devices and running Linux: @LittleCannon, @gabolander, @Daebis, @lbeltrame, @KishCom, @Xeraster, @colourful-land, @mauricioduarte01, @rogerxxxx, @winiciuscota

hello, could you make a tuto step by step to use yout patch ?
i'm on ubuntu

**Reply by nirenjan** [reply]
The fastest would be to use the existing CI scripts to build the sources.

1. Clone the repository - `git clone https://github.com/nirenjan/x52pro-linux.git -b mouse-support`
2. Change directory to the clone - `cd x52pro-linux`
3. Install the dependencies - `./.github/scripts/install-dependencies-ubuntu.sh`
4. Generate the build scripts - `./autogen.sh`
5. Build the changes - `mkdir build; cd build; ../configure --localstatedir=/var --sysconfdir=/etc && make -j`

The built files will be in the current directory after you have finished step 4. Then you can run `sudo ./x52d -vvvv -c ../daemon/x52d.conf`. This will start the daemon in background mode, with the PID in `/var/run/x52d.pid` and the log file in `/var/log/x52d.log`, using the configuration file from the sources. At this point, you can test out the mouse motion by moving the thumbstick, and you should see the mouse moving on screen.

You can also edit the configuration file, change the parameters, and trigger the daemon to reload the configuration by sending it the SIGHUP signal - `kill -HUP $(cat /var/run/x52d.pid)`

Let me know if you have any questions.

**Reply by Fr-Dae** [reply]
`sudo ./x52d -vvvv -c /dev/null`
![eyes](https://c.tenor.com/fpWpI-KaO5wAAAAC/spongebob-squarepants-my-eyes.gif)

**Reply by Fr-Dae** [reply]
*cracking fingers*
ok 


ok let's do the tour.

with the step-by-step help, it actually works, and it's pretty simple, since you just have to copy and paste.

but (and this is where there is room for improvement)

- no interface, even simplistic would be welcome. 
- no configuration possible (or I didn't understand how)
- all glows green and with full power 
- GMT +US time (am/pm) strange,

you don't want to do an interface?
ok let's do without, 
at the first launch of the app, a config file is created.
(and at the next launch it detects if the script exists if not it creates one)

and on the interface of the thruster of the questions

> Hello!
> BOnjour 
> Gunden Tag

select the right line and validate with right scroll.
(I will help you with the translation)

Time zone ? (with the roulette you can scroll the +/- gmt
12/24 ?
Date format ?
> yy-mm-dd
>mm-dd-yy
> dd-mm-yy

MFD light lvl ? <0-10>
LED level <0-10> ?
(flash led 1)
this button "name of button" 
> off
> green
> amber
> red

(etc for all button)


and at the end :
"thanks you for time. "


**mouse stick**
otherwise, concerning the configuration of the mouse...
The mousewhel is good, but the mouse movement, at least for office use, is awful and wird.
I'm not sure if it's a good idea to use the mouse in the same way as a mouse.

I'm not sure what to do with the mouse, but I'm sure it's a good idea to use it.
moving a few seconds causes a quick movement. (and thus offers precision in the action)

antimicroX offers a similar function.

I personally use a GX52 combo from roberto@leinardi.com
and AntimicroX https://github.com/AntiMicroX/antimicrox-profiles/tree/master/applications/StarConflict/03000000a30600006207000011010000 
for my games
Obviously it's not perfect, but it's better than nothing.

**Reply by Fr-Dae** [reply]
not perfect exemple ? with antimicroX on Starconflict i can't config my hotas "Z/S" like  i would.

it' +1 / 0 / -1 
not +100 / 0 / -25

:/  forget any notion of subtlety or softness in acceleration.

**Reply by Fr-Dae** [reply]
Lubuntu 18.04.5x64 (uptodate)

>`cat /proc/version && dpkg -l | grep -Ei "linux-headers|linux-image"`
`Linux version 4.15.0-156-generic (buildd@lcy01-amd64-008) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #163-Ubuntu SMP Thu Aug 19 23:31:58 UTC 2021
`

>`lsusb | grep X52: `
`Bus 001 Device 006: ID 06a3:0762 Saitek PLC Saitek X52 Pro Flight Control System`

**Reply by nirenjan** [reply]
Thank you so much for your detailed report. That's actually a great idea, but I have some other features that I need to implement before I can implement your idea.

With that said, there is a configuration file for the daemon. I left it out in the invocation since it would just use the defaults. However, if you want to edit the configuration, then you can use `daemon/x52d.conf` in the source tree. Also, you can change the invocation to `sudo ./x52d -vvvv -c ../daemon/x52d.conf`

You can also change the config file while the daemon is running, and reload the configuration by sending it the SIGHUP signal. You can use `pkill -HUP x52d`.

**Reply by nirenjan** [reply]
> not perfect exemple ? with antimicroX on Starconflict i can't config my hotas "Z/S" like  i would.
> 
> it' +1 / 0 / -1 
> not +100 / 0 / -25
> 
> :/  forget any notion of subtlety or softness in acceleration. 

Sorry, I don't understand this. I haven't used AntimicroX myself, so I'm not sure what you mean when you refer to "hotas Z/S".

**Reply by Fr-Dae** [reply]
azerty => zs it's 

> > not perfect exemple ? with antimicroX on Starconflict i can't config my hotas "Z/S" like  i would.
> > it' +1 / 0 / -1
> > not +100 / 0 / -25
> > :/  forget any notion of subtlety or softness in acceleration.
> 
> Sorry, I don't understand this. I haven't used AntimicroX myself, so I'm not sure what you mean when you refer to "hotas Z/S".

These are the keyboard keys for forward and backward. 
on the image above, (the key memo)

I use the throttle lever, for acceleration
so for forward, stop or reverse.

With a more detailed throttle management, I could manage the power given to the engines, for example to maintain me at a speed of 50 aulieux of my max which is 120.

because normally the throttle is a value between 0 and +100 and 0 and -25
except that with antimicroX, for the moment I have only managed "front" = z, "rear" = s
like on the keyboard, so not much subtlety in acceleration, or manoeuvre.

**Reply by nirenjan** [reply]
I don't know of any application that can keep the throttle at a certain value with only keyboard input. It seems to me that either the game needs to support joystick inputs directly, or lacking that, some sort of a plugin that would interact with both the game and the translation tool (AntimicroX or similar), and dynamically decide to press a set of keys or not.

**Reply by Fr-Dae** [reply]
StarConflict has no "real" hotas compatibility and it's junk code x)

**Comment by Fr-Dae** [comment]
https://i.imgur.com/M18bJfu.jpg
you talking about "mousse3" stick ?

**Reply by nirenjan** [reply]
Yes, the scroll wheel on the back, the little "Push to Talk" button and the thumbstick next to it. The thumbstick controls the mouse movement, while the Push to Talk button is the Left mouse button, and pressing the scroll wheel is the Right mouse button.

**Reply by Fr-Dae** [reply]
ok, i will try this, but that can maybe take some weeks, my ssd are very full.

**Reply by nirenjan** [reply]
No problem, I appreciate any help you can provide.

**Reply by Fr-Dae** [reply]
INSTALL.md
lign 29 sudo apt-get ===> aftr 2014 you need just sudo apt, no more -get

> cd x52pro-linux
you should say "cd ./x52pro-linux"

> mkdir build; cd build; ../configure --localstatedir=/var --sysconfdir=/etc && make -j
bash: ../configure: Aucun fichier ou dossier de ce type


:/

**Reply by Fr-Dae** [reply]
https://controlc.com/c6fa2dcf

**Reply by Fr-Dae** [reply]
bash: ../configure.ac: Permission non accordée

i think you have forget make  "sudo chmod -R 775 ./" 
on ./x52pro-linux/ folder

**Reply by Fr-Dae** [reply]
```
dae@dae-GL552V:~/Games/HOTAS/x52pro-linux$ ./configure.ac --localstatedir=/var --sysconfdir=/etc && make -j
./configure.ac: ligne 7: erreur de syntaxe près du symbole inattendu « [x52pro-linux], »
./configure.ac: ligne 7: `AC_INIT([x52pro-linux], [0.2.2], [nirenjan@gmail.com])'
```

**Reply by nirenjan** [reply]
It seems like I forgot a step between 3 & 4. You got part of it - `cd x52pro-linux`. You then need to run `./autogen.sh` to generate the configure file. Then proceed with the instructions in step 4.

Thank you for pointing it out.

**Reply by Fr-Dae** [reply]
ok, edit your post https://github.com/nirenjan/x52pro-linux/discussions/34#discussioncomment-1335685
for the next guys , who would come and get lost here ;)

**Comment by Fr-Dae** [comment]
I wanted to know, why not work in collaboration with leinardi. who already has a functional base (and interface) but does not have the time to maintain his repo.

and you who have well advanced on the basic function of the buttons/manipulators. but don't have an interface. 

You would both probably benefit from it



Dae#5125 on discord (for me)
and https://discord.gg/HTCrmU8 for leinardi's discord

**Reply by nirenjan** [reply]
I felt that the licenses were incompatible, his repo is using GPLv3, and I do not wish to upgrade from GPLv2 for my own ideological reasons. That being said, I do have a plan to get a GUI in eventually, and I have already started working towards that goal, it's just that life and my day job take precedence over this hobby project.

**Reply by Fr-Dae** [reply]
> I felt that the licenses were incompatible, his repo is using GPLv3, and I do not wish to upgrade from GPLv2 for my own ideological reasons. That being said, I do have a plan to get a GUI in eventually, and I have already started working towards that goal, it's just that life and my day job take precedence over this hobby project.

 you can just use CC-BY-NC-SA x) and problems solve
